### PR TITLE
[Feature] SLEM speculative decoding for heterogeneous vocabularies

### LINKED ISSUE #45020: [Feature Request] SLEM (String-Level Exact Match) speculative decoding for heterogeneous vocabularies
## Summary

Implement **SLEM (String-Level Exact Match)** speculative decoding as a complement to TLI (#38173 / #38174). While TLI restricts the draft model to the vocabulary intersection, SLEM lets the draft use its **full vocabulary** and verifies at the text level — achieving higher acceptance rates when vocabulary overlap is low (cross-family pairs).

Both algorithms are from the same ICML 2025 oral paper:
> *Accelerating LLM Inference with Lossless Speculative Decoding Algorithms for Heterogeneous Vocabularies*
> Timor et al., https://arxiv.org/abs/2502.05202

## Motivation

TLI (PR #38174) works well for same-family pairs with 95%+ vocabulary overlap (e.g., Qwen2.5-0.5B → Qwen2.5-1.5B). But for cross-family pairs — which are the most common real-world scenario — the intersection shrinks significantly:

| Draft → Target | Overlap |
|----------------|---------|
| Qwen2.5-0.5B → Qwen2.5-1.5B | 99.8% |
| SmolLM2-135M → Qwen2.5-1.5B | ~72% |
| Llama-3.2-1B → Qwen2.5-7B | ~75% |

When TLI constrains draft logits to a 72% intersection, the draft model's quality degrades — it can't express ~28% of its vocabulary, leading to lower acceptance rates and reduced speedup.

SLEM solves this by operating at the **string level**: the draft generates freely, its output is decoded to text and re-tokenized for the target, and verification uses exact matching. The draft's full expressiveness is preserved.

## Algorithm (from the paper)

1. Draft model generates `K` tokens using its **full vocabulary** (no logit masking).
2. Decode draft tokens to text, re-tokenize with target tokenizer → `m` target tokens.
3. Single batched forward pass on target model over the `m` candidate tokens.
4. Sample from target logits at each position: `t'_1, ..., t'_{m+1}`.
5. Find first mismatch: `j = argmin_j { t'_j != t_j }`.
6. Accept `t_1, ..., t_{j-1}, t'_j` (matched prefix + bonus token from target).

**Key properties:**
- Provably lossless (Theorem 2.1 in the paper)
- Draft uses full vocabulary → higher quality proposals
- Up to 2.8x speedup reported in the paper (vs. 1.7x for TLI)

## References

- Paper: https://arxiv.org/abs/2502.05202
- HuggingFace Transformers implementation (merged): https://github.com/huggingface/transformers/pull/35029
- TLI in vLLM: #38174
- TLI in SGLang: https://github.com/sgl-project/sglang/pull/22883

### PR DESCRIPTION
**Pull request by jmamou**
## Summary

Implements **SLEM (String-Level Exact Match)** speculative decoding — a text-level verification algorithm that lets the draft model use its **full vocabulary** when draft and target have different tokenizers.

Closes #45020

Based on the ICML 2025 oral paper:
> *Accelerating LLM Inference with Lossless Speculative Decoding Algorithms for Heterogeneous Vocabularies*
> Nadav Timor, Jonathan Mamou, Daniel Korat, Moshe Berchansky, Gaurav Jain, Oren Pereg, Moshe Wasserblat, David Harel
> https://arxiv.org/abs/2502.05202

SLEM is **provably lossless** — the target distribution is never modified.

## Algorithm

**SLEM (String-Level Exact Match):**
1. Draft model generates K tokens using its **full vocabulary** (no logit masking).
2. Decode draft tokens to text, re-tokenize with target tokenizer → candidate target tokens.
3. Target model scores the candidates in a single batched forward pass.
4. Greedy verification: accept matched prefix + bonus token.

## When to use SLEM vs TLI

| Scenario | TLI (#38174) | SLEM (this PR) |
|----------|-----|------|
| Same family (99%+ overlap) | ✅ simple, fast | works but unnecessary overhead |
| Cross-family (70-90% overlap) | draft quality degraded | ✅ full vocab → higher acceptance |
| Byte-level / char-level draft | tiny intersection | ✅ works naturally |

## Changes

| File | Change |
|------|--------|
| `vllm/v1/spec_decode/slem.py` | `SlemMapper` — string-level decode + re-encode for lossless token mapping |
| `vllm/v1/spec_decode/llm_base_proposer.py` | Integrate SLEM into sampling and propose loop |
| `vllm/v1/spec_decode/draft_model.py` | Load dual tokenizers, instantiate mapper |
| `vllm/config/speculative.py` | Add `use_heterogeneous_vocab` + `heterogeneous_vocab_method` |
| `tests/v1/spec_decode/test_slem.py` | Unit tests for SLEM mapper |
| `tests/v1/e2e/spec_decode/test_slem_e2e.py` | E2E losslessness test |

## Usage

```bash
vllm serve microsoft/phi-4 \
    --speculative-config '{"method": "draft_model", "model": "Qwen/Qwen2.5-0.5B-Instruct", "num_speculative_tokens": 5, "use_heterogeneous_vocab": true, "heterogeneous_vocab_method": "slem"}'
```

## Benchmark Results

| Target | Draft | Baseline | SLEM | Speedup |
|--------|-------|----------|------|---------|
| microsoft/phi-4 | Qwen/Qwen2.5-0.5B-Instruct | 14.05s | 7.44s | **1.89x** |
| microsoft/phi-4 | Qwen/Qwen3-0.6B | 13.93s | 7.72s | **1.80x** |

*RTX 6000 Ada, long-context CNN/DailyMail prompts, 512 max output tokens, greedy decoding.*

### Vocabulary Analysis

| Pair | Target Vocab | Draft Vocab | String Overlap | 1:1 ID Mapping |
|------|-------------|-------------|----------------|----------------|
| phi-4 + Qwen2.5-0.5B-Instruct | 100,352 | 151,665 | 98.8% | 65.7% |
| phi-4 + Qwen3-0.6B | 100,352 | 151,669 | 98.8% | ~65.7% |

**String overlap** = fraction of target vocab token strings that also exist in draft vocab.
**1:1 ID mapping** = fraction of draft tokens where decode→re-encode produces exactly one target token (what a per-token lookup requires). The remaining 34.3% need SLEM’s string-level mapping.

## Test Plan

- [x] Unit tests pass: `tests/v1/spec_decode/test_slem.py` (8 tests)
- [x] E2E losslessness test: `tests/v1/e2e/spec_decode/test_slem_e2e.py`
- [x] Benchmark: 1.89x speedup on phi-4 + Qwen2.5-0.5B cross-family pair

## Related

- Paper: https://arxiv.org/abs/2502.05202
- HuggingFace Transformers implementation: https://github.com/huggingface/transformers/pull/35029
- vLLM TLI PR: #38174
- SGLang TLI: https://github.com/sgl-project/sglang/pull/22883

---
<details>
<summary>Essential Elements of an Effective PR Description Checklist</summary>

- [x] The purpose of the PR
- [x] The test plan
- [x] The test results
- [ ] Documentation update
</details>

### CODE REVIEW

**Comment by danielkorat** [comment]
hey @benchislett @luccafong @MatthewBonanni — rebased this onto main and fixed conflicts (kept TLI working, added SLEM as an alterate method). would appreciate a review when you get a chance, CI's queued pending the pre-run-check gate.

**Comment by MLKoz2** [comment]
Very useful feature, nevertheless more efficiency comparisons would be useful, TLI vs SLEM for various pairs of models.
